pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -356,16 +348,21 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 					continue
 				}
 
-				for _, cn := range con.Clusters {
+				previous := sets.NewSet(con.Clusters...)
+				current := sets.NewSet(clusters...)
+
+				// removed clusters
+				for cn := range previous.Difference(current) {
 					s.removeEdsCon(cn, con.ConID)
 				}
 
-				for _, cn := range clusters {
-					s.getOrAddEdsCluster(cn, con.ConID, con)
+				// new added clusters
+				for cn := range current.Difference(previous) {
+					s.getOrAddEdsCluster(cn, con)
 				}
```

Comment:
@howardjohn @ramaraochavali This is the purpose, reduce the remove/add ops

Reply:
Ok. Makes sense

Reply:
@hzxuzhonghu Would it help if we add a new method to `DiscoveryServer` called `updateEdsClusters(addedclusters, removedclusters)` and pass all clusters at one shot to it and update it? One benefit I can see is we can take lock once and update it - rather than taking multiple locks,. WDYT?